api/src/play_asset_pack_request_tracker.gd

```diff
@@ -0,0 +1,65 @@
+# ##############################################################################
+#
+#	Copyright 2020 Google LLC
+#
+#	Licensed under the Apache License, Version 2.0 (the "License");
+#	you may not use this file except in compliance with the License.
+#	You may obtain a copy of the License at
+#
+#		https://www.apache.org/licenses/LICENSE-2.0
+#
+#	Unless required by applicable law or agreed to in writing, software
+#	distributed under the License is distributed on an "AS IS" BASIS,
+#	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#	See the License for the specific language governing permissions and
+#	limitations under the License.
+#
+# ##############################################################################
+#
+# The PlayAssetPackRequestTracker class generates unique signal_id integers and
+# keeps a mapping of signal_id to Request objects. This signal_id is used as an
+# identifier and passed to the Android plugin. Eventually the Android plugin will
+# emit signals along with this signal_id. In this way we can know which signal 
+# emitted from the plugin corresponds to which Request object
+#
+# ##############################################################################
+class_name PlayAssetPackRequestTracker
+extends Object
+
+const _SIGNAL_ID_MIN = 0
+
+var _signal_id_counter : int
+var _register_request_mutex : Mutex
+var _signal_id_to_request_map : Dictionary
+
+func _init():
+	_signal_id_counter = _SIGNAL_ID_MIN
+	_signal_id_to_request_map = Dictionary()
+	_register_request_mutex = Mutex.new()
+
+func get_current_signal_id() -> int:
+	return _signal_id_counter
+
+func _increment_signal_id() -> void:
+	_signal_id_counter += 1
+
+# registers the request object and returns the signal_id assigned
+func register_request(request : PlayAssetPackRequest) -> int:
```

Comment:
Are these methods going to be called by developers or only by PlayAssetDeliveryManager? If developers aren't going to call these methods directly, it might be better to move the mutex to PlayAssetDeliveryManager and lock the entire `play_asset_pack_request_tracker` object when you call its methods.

Reply:
i.e. Don't make this class thread-safe but make PlayAssetDeliveryManager thread-safe and give it the responsibility of locking this object.

Reply:
I'm think about this approach. However, one of the tradeoffs for this is that it can prevent us to write concurrency related unit test cases testing only the PlayAssetPackRequestTracker class. Do you think if there is any other solutions for this?

Reply:
Also since we will be calling ```register_request()``` multiple times, we will need to write ```lock()``` and ```unlock()``` statements multiple times if they are refactored out of ```register_request()```. This can create repeated code blocks in PlayAssetDeliveryManager.